genvm/code_enforcement.py

```diff
@@ -0,0 +1,165 @@
+import ast
+
+def code_enforcement_check(code: str, class_name:str) -> str:
+    result = {"status": "error", "message": "", "data": []}
+    # Check is valid Python code
+    code_check_result = _check_code(code)
+    if code_check_result:
+        result["message"] = code_check_result
+        return result
+    # See if the class exists
+    if not _does_class_exist_in_code(code, class_name):
+        result["message"] = f"The class {class_name} does not exist in the code"
+        return result
+    # Make sure there are no raw instantiations of the EquivalencePrinciple class
+    if _code_has_bad_implementations_of_eq_principle(code):
+        result["message"] = "You cannot directly instantiate the EquivalencePrinciple class"
+        return result
+    # Make sure that no code modifies self inside an equivalence block
+    linenos_list = _linenos_of_where_code_modifys_self_in_eq_block(code,class_name)
+    if len(linenos_list):
+        result["message"] = "Self was modified inside an equivalence block"
+        result["data"] = linenos_list
+        return result
+    # Make sure no equivalence block variables are referenced outside the block
+    linenos_list = _code_references_eq_block_variables(code, class_name)
+    if len(linenos_list):
+        result["message"] = "Variables declared in the equivalence block are referenced outside of the equivalence block"
+        result["data"] = linenos_list
+        return result
+    result["status"] = "success"
+    return result
+
+def _check_code(code:str):
+    try:
+        ast.parse(code)
+        return None
+    except Exception:
+        return "Your code is not valid Python code"
+
+
+def _does_class_exist_in_code(code:str, class_name:str) -> bool:
+    visitor = ClassExistsVisitor(class_name)
+    visitor.visit(ast.parse(code))
+    return visitor.class_exists
+
+
+class ClassExistsVisitor(ast.NodeVisitor):
+    def __init__(self, class_name):
+        self.class_name = class_name
+        self.class_exists = False
+
+    def visit_ClassDef(self, node):
+        if node.name == self.class_name:
+            self.class_exists = True
+        self.generic_visit(node)
```

Comment:
@sevenearths please add some comments or explain how this approach works.

Reply:
The file has been updated accordingly => https://github.com/yeagerai/genlayer-simulator/pull/146/commits/529a9b991a0f8be0df5b1bc9ba50b55832cd5efe